parse/ShipDesignsParser.cpp

```diff
@@ -5,24 +5,65 @@
 #include "../universe/ShipDesign.h"
 
 #include <boost/spirit/include/phoenix.hpp>
-
+#include <boost/phoenix/function/adapt_function.hpp>
+#include <boost/uuid/nil_generator.hpp>
+#include <boost/uuid/uuid_io.hpp>
 
 FO_COMMON_API extern const int ALL_EMPIRES;
 
 #define DEBUG_PARSERS 0
 
 #if DEBUG_PARSERS
 namespace std {
-    inline ostream& operator<<(ostream& os, const std::map<std::string, ShipDesign*>&) { return os; }
-    inline ostream& operator<<(ostream& os, const std::pair<const std::string, ShipDesign*>&) { return os; }
+    inline ostream& operator<<(ostream& os, const std::map<std::string, std::unique_ptr<ShipDesign>>&) { return os; }
+    inline ostream& operator<<(ostream& os, const std::pair<const std::string, std::unique_ptr<ShipDesign>>&) { return os; }
     inline ostream& operator<<(ostream& os, const std::vector<std::string>&) { return os; }
 }
 #endif
 
 namespace {
     const boost::phoenix::function<parse::detail::is_unique> is_unique_;
 
-    const boost::phoenix::function<parse::detail::insert> insert_;
+    void insert_ship_design(std::map<std::string, std::unique_ptr<ShipDesign>>& designs,
+                            const std::string& name, const std::string& description,
+                            const std::string& hull, const std::vector<std::string>& parts,
+                            const std::string& icon, const std::string& model,
+                            bool name_desc_in_stringtable, const boost::uuids::uuid& uuid)
+    {
+        // TODO use make_unique when converting to C++14
+        auto design = std::unique_ptr<ShipDesign>(
+            new ShipDesign(name, description, 0, ALL_EMPIRES, hull, parts, icon, model, name_desc_in_stringtable, false, uuid));
+
+        auto inserted_design = designs.insert(std::make_pair(design->Name(false), std::move(design)));
+    };
+    BOOST_PHOENIX_ADAPT_FUNCTION(void, insert_ship_design_, insert_ship_design, 9)
+
+    // A UUID validator
+    bool is_valid_uuid(const std::string& uuid_string) {
+        try {
+            boost::lexical_cast<boost::uuids::uuid>(uuid_string);
+        } catch (boost::bad_lexical_cast&) {
+            ErrorLogger() << uuid_string << " is not a valid UUID.  A valid UUID looks like 01234567-89ab-cdef-0123-456789abcdef";
+            return false;
+        }
+        return true;
+    };
+
+    // A lazy UUID parser
+    BOOST_PHOENIX_ADAPT_FUNCTION(bool, is_valid_uuid_, is_valid_uuid, 1)
+
+    // A UUID parser
+    boost::uuids::uuid parse_uuid(const std::string& uuid_string) {
+        try {
+            return boost::lexical_cast<boost::uuids::uuid>(uuid_string);
+        } catch (boost::bad_lexical_cast&) {
+            // This should never happen because the previous is_valid should short circuit.
+            ErrorLogger() << uuid_string << " is not a valid UUID.  A valid UUID looks like 01234567-89ab-cdef-0123-456789abcdef";
+            return boost::uuids::uuid{{0}};
```

Comment:
They are not distinct.  

I should be using the `nil_generator()` or `nil_uuid()` and the test `is_nil()` everywhere, but I started out using the constructor `{{0}}` and an equality test.

Eventually, all of the nil generators are/should be replaced with random generators.  `{{0}}` is like the null pointer of UUID.  It indicates that something is not initialized.

I've removed the `{{0}}` from this PR.